src/group/managed_group/mod.rs

```diff
@@ -123,60 +201,221 @@ impl ManagedGroup {
+                    LeafIndex::from(*member),
+                )
+            })
+            .collect();
+
+        let mls_messages = self.plaintext_to_mls_messages(plaintext_messages);
+
+        Ok(mls_messages)
+    }
+
+    /// Leave the group
+    pub fn leave_group(
+        &mut self,
+        credential_bundle: &CredentialBundle,
+    ) -> Result<Vec<MLSMessage>, ManagedGroupError> {
+        let remove_proposal = self.group.create_remove_proposal(
+            &self.aad,
+            &credential_bundle,
+            LeafIndex::from(self.group.tree().get_own_node_index()),
+        );
+
+        let mls_messages = self.plaintext_to_mls_messages(vec![remove_proposal]);
+
+        Ok(mls_messages)
     }
 
     /// Gets the current list of members
-    pub fn get_members(&self) -> &[Credential] {
-        unimplemented!()
+    pub fn get_members(&self) -> Vec<Credential> {
+        let mut members: Vec<Credential> = vec![];
+        let tree = self.group.tree();
+        let leaf_count = self.group.tree().leaf_count();
+        for index in 0..leaf_count.as_usize() {
+            let leaf = &tree.nodes[LeafIndex::from(index)];
+            if let Some(leaf_node) = leaf.get_key_package_ref() {
+                members.push(leaf_node.credential().clone());
+            }
+        }
+        members
     }
 
     // === Process messages ===
 
     /// Processes any incoming messages from the DS (MLSPlaintext &
     /// MLSCiphertext) and triggers the corresponding callback functions
     pub fn process_messages(&mut self, messages: &[MLSMessage]) {
-        unimplemented!()
+        // Iterate over all incoming messages
+        for message in messages {
+            // Check the type of message we recived
+            let (plaintext_option, aad_option) = match message {
+                // If it is a ciphertext we decrypt it and return the plaintext by value
+                MLSMessage::Ciphertext(ciphertext) => {
+                    let aad = &ciphertext.authenticated_data;
+                    match self.group.decrypt(ciphertext) {
+                        Ok(plaintext) => (Some(ValueOrRef::from(plaintext)), Some(aad)),
+                        Err(_) => {
+                            // If there is a callback for that event we should call it
+                            self.invalid_message_event(InvalidMessageError::InvalidCiphertext(
+                                aad.to_vec(),
+                            ));
+                            (None, Some(aad))
+                        }
+                    }
+                }
+                // If it is a plaintext message we just return the reference
+                MLSMessage::Plaintext(plaintext) => (Some(ValueOrRef::from(plaintext)), None),
+            };
+            // If it was a plaintext message or if the decryption succeeded we continue,
+            // otherwise we move to the next message
+            if let Some(plaintext) = plaintext_option {
```

Comment:
If not, we don't do anything? Would be nice to let the caller know that something didn't work out.

Reply:
We already throw an error earlier if the decryption failed.

Reply:
Why not? The decryption might still fail, in which case we get a `None`.

Reply:
You said you're throwing an error when the decryption fails. So either we throw an error and `plaintext` is always available here, or `plaintext` is an option and and we have to handle the `None` case.

Reply:
By throwing an error I meant that the corresponding callback function gets called, but it doesn't halt the processing of messages altogether. We just move to the next message in the queue instead.
I guess I can refactor the code to use something like `continue` instead, but I personally don't like these construction because they are hard to understand and review.

Reply:
If you have trouble with `continue`, you can use labelled continues in Rust to make your life a little easier.
The issue is that there's an `if` now without an `else` branch, which makes me itchy.